structuralcodes/codes/mc2020/_concrete_limit_state_of_cracking.py

```diff
@@ -0,0 +1,1367 @@
+    return max(
+        (sigma_s_x - beta_TS_x * sigma_sr_ef_x) / Es,
+        sigma_s_x / Es * (1 - beta_TS_x),
+    )
+
+
+def eps_sm_y_eps_cm_y(
+    sigma_s_y: float, sigma_sr_ef_y: float, Es: float, beta_TS_y: float
+) -> float:
+    """Calculate the relative mean strain (eps_sm,y - eps_cm,y) in the y
+    direction.
+
+    fib Model Code 2020, eq. (30.5-21)
+
+    Args:
+        sigma_s_y (float): Reinforcing steel stress at the crack in y direction
+            in MPa.
+        sigma_sr_ef_y (float): Steel stress at the crack formation stage in y
+            direction in MPa.
+        Es (float): Modulus of elasticity of steel in MPa.
+        beta_TS_y (float): Empirical coefficient for y direction.
+
+    Returns:
+        float: Relative mean strain (eps_sm,y - eps_cm,y).
+
+    Raises:
+        ValueError: If any input is invalid.
+    """
+    if (
+        sigma_s_y < 0
+        or sigma_sr_ef_y < 0
+        or Es <= 0
+        or not (0 <= beta_TS_y <= 1)
+    ):
+        raise ValueError(
+            'Invalid input values for relative mean '
+            + 'strain calculation in y direction.'
+        )
+
+    return max(
+        (sigma_s_y - beta_TS_y * sigma_sr_ef_y) / Es,
+        sigma_s_y / Es * (1 - beta_TS_y),
+    )
+
+
+def eps_2(tau_xy: float, Ec: float, theta: float) -> float:
+    """Calculate the principal compressive strain eps_2.
+
+    fib Model Code 2020, eq. (30.5-19)
+
+    Args:
+        tau_xy (float): Shear stress in MPa.
+        Ec (float): Modulus of elasticity of concrete in MPa.
+        theta (float): Angle between the x direction reinforcement and the
+            direction of principal tensile stress in degrees.
+
+    Returns:
+        float: Principal compressive strain ε2.
+
+    Raises:
+        ValueError: If any input is invalid.
+    """
+    if tau_xy < 0 or Ec <= 0 or not (0 <= theta <= 90):
+        raise ValueError(
+            'Invalid input values for principal '
+            + 'compressive strain calculation.'
+        )
+
+    theta_rad = math.radians(theta)
+    sin_theta = math.sin(theta_rad)
+    cos_theta = math.cos(theta_rad)
+
+    return -abs(tau_xy) / (Ec * sin_theta * cos_theta)
+
+
+def sigma_s_x(
+    sigma_x: float, tau_xy: float, rho_s_ef_x: float, theta: float
+) -> float:
+    """Calculate the reinforcing steel stress sigma_s,x at the crack in the x
+    direction.
+
+    fib Model Code 2020, eq. (30.5-21)
+
+    Args:
+        sigma_x (float): Normal stress in x direction in MPa.
+        tau_xy (float): Shear stress in MPa.
+        rho_s_ef_x (float): Effective reinforcement ratio in the y direction.
+        theta (float): Angle between the y direction reinforcement and the
+            direction of principal tensile stress in degrees.
+
+    Returns:
+        float: Steel stress sigma_s,y in the x direction in MPa.
+
+    Raises:
+        ValueError: If any input is invalid.
+    """
+    if sigma_x < 0 or tau_xy < 0 or rho_s_ef_x <= 0 or not (0 <= theta <= 90):
```

Comment:
As previous comment: are we sure both siga_x, sigma_y and tau_xy must be greater than zero? Can one component (sigma_x or sigma_y) be exactly equal to zero?